# 1. Toronto neighborhoods overview

I will create a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name.

Firstly I will download necessary packages

In [1]:
from bs4 import BeautifulSoup
import requests
import pandas as pd

Getting the data from Wikipedia

In [2]:
List_url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
source = requests.get(List_url).text
soup = BeautifulSoup(source, 'xml')
table_contents=[]
table=soup.find('table')

Scraping the data according to [hints for scraping Notebook](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/NewLinkWebscrapingHints.md)

In [3]:
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

# renaming PostalCode column
df.rename(columns={'PostalCode':'Postal Code'}, inplace=True)

df.head()

Borough                      Neighborhood Postal Code
0        North York                         Parkwoods         M3A
1        North York                  Victoria Village         M4A
2  Downtown Toronto         Regent Park, Harbourfront         M5A
3        North York  Lawrence Manor, Lawrence Heights         M6A
4      Queen's Park     Ontario Provincial Government         M7A

Re-grouping according to the table in assignement

In [4]:
df = df.groupby(['Postal Code','Borough'])['Neighborhood'].apply(', '.join).reset_index()
df.head(10)

Postal Code      Borough                                     Neighborhood
0         M1B  Scarborough                                   Malvern, Rouge
1         M1C  Scarborough           Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough                Guildwood, Morningside, West Hill
3         M1G  Scarborough                                           Woburn
4         M1H  Scarborough                                        Cedarbrae
5         M1J  Scarborough                              Scarborough Village
6         M1K  Scarborough      Kennedy Park, Ionview, East Birchmount Park
7         M1L  Scarborough                  Golden Mile, Clairlea, Oakridge
8         M1M  Scarborough  Cliffside, Cliffcrest, Scarborough Village West
9         M1N  Scarborough                      Birch Cliff, Cliffside West

Let's see the number of rows:

In [5]:
df.shape

(103, 3)

So we have 103 rows in our dataframe.

# 2. Location data of Toronto neighborhoods

Let's download csv file with geo location data.

In [6]:
df1=pd.read_csv('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv')
df1.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Now we combine both dataframes (neighborhoods and geo location data) into one dataframe named geoData.

In [7]:
geoData = pd.merge(df, df1, on='Postal Code')
geoData.head()

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

# 3. Segmenting and Clustering

Let's install libraries first.

In [8]:
import numpy as np

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json

!pip install geopy --user

from geopy.geocoders import Nominatim 

from pandas.io.json import json_normalize 

import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

!pip install folium --user
import folium 

#### Get the latitude and longitude values of Toronto, Canada

In [9]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Create a map of Toronto with neighborhoods superimposed on top.

In [10]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(geoData['Latitude'], geoData['Longitude'], geoData['Borough'], geoData['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

 #### Filtering boroughs that contain the word Toronto

I will filter and visualize only boroughs which contain word Toronto.

In [11]:
df_toronto = geoData[geoData['Borough'].str.contains("Toronto")].reset_index(drop=True)
df_toronto.head()

Postal Code                 Borough                    Neighborhood  \
0         M4E            East Toronto                     The Beaches   
1         M4J  East York/East Toronto              The Danforth  East   
2         M4K            East Toronto    The Danforth West, Riverdale   
3         M4L            East Toronto  India Bazaar, The Beaches West   
4         M4M            East Toronto                 Studio District   

    Latitude  Longitude  
0  43.676357 -79.293031  
1  43.685347 -79.338106  
2  43.679557 -79.352188  
3  43.668999 -79.315572  
4  43.659526 -79.340923

In [12]:
map_toronto_only = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto_only)  
    
map_toronto_only

#### Define Foursquare Credentials and Version

Now, let's get the top 100 venues that are in boroughs which contain word Toronto within a radius of 500 meters.

In [13]:
CLIENT_ID = 'SDOSSXMTMQSAXSIXGNMEQO1Y4ZB3XXRZ5YLAVARXTPY13QW3' 
CLIENT_SECRET = '2QZVKTERDIARRLNRMV31Q1BH0TMR1C1D1H3UEPG4PFJKYUSY' 
VERSION = '20210407' 
LIMIT = 100 

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: SDOSSXMTMQSAXSIXGNMEQO1Y4ZB3XXRZ5YLAVARXTPY13QW3
CLIENT_SECRET:2QZVKTERDIARRLNRMV31Q1BH0TMR1C1D1H3UEPG4PFJKYUSY


Explore Neighborhoods in boroughs containing word Toronto

In [14]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now we create a new dataframe called _toronto_venues_.

In [15]:
toronto_venues = getNearbyVenues(names=df_toronto['Neighborhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )

The Beaches
The Danforth  East
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Enclave of M5E
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High Park, The Junction 

Let's have a look on _toronto_venues_ dataframe 

In [16]:
print(toronto_venues.shape)
toronto_venues.head()

(1579, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         The Beaches              43.676357              -79.293031   
1         The Beaches              43.676357              -79.293031   
2         The Beaches              43.676357              -79.293031   
3         The Beaches              43.676357              -79.293031   
4  The Danforth  East              43.685347              -79.338106   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                       Upper Beaches       43.680563       -79.292869   
4                    Danforth & Jones       43.684352       -79.334792   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3       Neighborhood  
4       Intersection

And group it by _Neighborhood_.

In [17]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            59   
Brockton, Parkdale Village, Exhibition Place                           24   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     64   
Christie                                                               16   
Church and Wellesley                                                   74   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        6   
Dufferin, Dovercourt Village                                           16   
Enclave of M4L                                                         17   
Enclave of M5E                                                         99   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          26   
India Bazaar, The Beaches West                                         21   
Kensington Market, Chinatown, Grange Park                              64   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               44   
Moore Park, Summerhill East                                             2   
North Toronto West                                                     19   
Parkdale, Roncesvalles                                                 15   
Regent Park, Harbourfront                                              44   
Richmond, Adelaide, King                                               93   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     33   
St. James Town                                                         80   
St. James Town, Cabbagetown                                            46   
Studio District                                                        36   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    19   
The Beaches                                                             4   
The Danforth  East                                                      4   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             59   
Brockton, Parkdale Village, Exhibition Place                            24   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      64   
Christie                                                                16   
Church and Wellesley                                                    74   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              33   
Davisville North                                                         6   
Dufferin, Dovercourt Village        

In [18]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 235 uniques categories.


#### Analyze of each neighborhood

In [19]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
neigh_column = toronto_onehot['Neighborhood']
toronto_onehot.drop(labels=["Neighborhood"], axis=1, inplace=True)
toronto_onehot.insert(0, 'Neighborhood', neigh_column)

toronto_onehot.head()

Neighborhood  Airport  Airport Food Court  Airport Gate  \
0         The Beaches        0                   0             0   
1         The Beaches        0                   0             0   
2         The Beaches        0                   0             0   
3         The Beaches        0                   0             0   
4  The Danforth  East        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       0     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               0        0                0         0             0   

   Burrito Place  Bus Line  Bus Stop  Business Service  Butcher  Café  \
0              0         0         0                 0        0     0   
1              0         0         0                 0        0     0   
2              0         0         0                 0        0     0   
3              0         0         0                 0        0     0   
4              0         0         0

And let's examine the new dataframe size.

In [20]:
toronto_onehot.shape

(1579, 235)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [21]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood   Airport  \
0                                         Berczy Park  0.000000   
1        Brockton, Parkdale Village, Exhibition Place  0.000000   
2   CN Tower, King and Spadina, Railway Lands, Har...  0.058824   
3                                  Central Bay Street  0.000000   
4                                            Christie  0.000000   
5                                Church and Wellesley  0.000000   
6                      Commerce Court, Victoria Hotel  0.000000   
7                                          Davisville  0.000000   
8                                    Davisville North  0.000000   
9                        Dufferin, Dovercourt Village  0.000000   
10                                     Enclave of M4L  0.000000   
11                                     Enclave of M5E  0.000000   
12             First Canadian Place, Underground city  0.000000   
13                           Forest Hill North & West  0.000000   
14                           Garden District, Ryerson  0.000000   
15  Harbourfront East, Union Station, Toronto Islands  0.000000   
16                      High Park, The Junction South  0.000000   
17                     India Bazaar, The Beaches West  0.000000   
18          Kensington Market, Chinatown, Grange Park  0.000000   
19                                      Lawrence Park  0.000000   
20                           Little Portugal, Trinity  0.000000   
21                        Moore Park, Summerhill East  0.000000   
22                                 North Toronto West  0.000000   
23                             Parkdale, Roncesvalles  0.000000   
24                          Regent Park, Harbourfront  0.000000   
25                           Richmond, Adelaide, King  0.000000   
26                                           Rosedale  0.000000   
27                                           Roselawn  0.000000   
28                                 Runnymede, Swansea  0.000000   
29                                     St. James Town  0.000000   
30                        St. James Town, Cabbagetown  0.000000   
31                                    Studio District  0.000000   
32  Summerhill West, Rathnelly, South Hill, Forest...  0.000000   
33                The Annex, North Midtown, Yorkville  0.000000   
34                                        The Beaches  0.000000   
35                                 The Danforth  East  0.000000   
36                       The Danforth West, Riverdale  0.000000   
37           Toronto Dominion Centre, Design Exchange  0.000000   
38                     University of Toronto, Harbord  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000         0.000000   
1             0.000000      0.000000        0.000000         0.000000   
2             0.058824      0.058824        0.117647         0.117647   
3             0.000000      0.000000        0.000000         0.000000   
4             0.000000      0.000000        0.000000         0.000000   
5             0.000000      0.000000        0.000000         0.000000   
6             0.000000      0.000000        0.000000         0.000000   
7             0.000000      0.000000        0.000000         0.000000   
8             0.000000      0.000000        0.000000         0.000000   
9             0.000000      0.000000        0.000000         0.000000   
10            0.000000      0.000000        0.000000         0.000000   
11            0.000000      0.000000        0.000000         0.000000   
12            0.000000      0.000000        0.000000         0.000000   
13            0.000000      0.000000        0.000000         0.000000   
14            0.000000      0.000000        0.000000         0.000000   
15            0.000000      0.000000        0.000000         0.000000   
16            0.000000      0.000000        0.000000         0.000000   
17            0.000000      0.000000        0.

Let's see the new size

In [22]:
toronto_grouped.shape

(39, 235)

#### Let's print each neighborhood along with the top 5 most common venues

In [23]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
          venue  freq
0   Coffee Shop  0.07
1  Cocktail Bar  0.05
2        Bakery  0.05
3    Restaurant  0.03
4      Pharmacy  0.03


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.12
1          Bakery  0.08
2     Coffee Shop  0.08
3  Breakfast Spot  0.08
4   Burrito Place  0.04


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.12
1   Airport Service  0.12
2  Airport Terminal  0.12
3           Airport  0.06
4               Bar  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.17
1                Café  0.06
2      Sandwich Place  0.06
3  Italian Restaurant  0.05
4         Salad Place  0.03


----Christie----
           venue  freq
0  Grocery Store  0.25
1           Café  0.19
2           Park  0.12
3     Baby Store  0.06
4    Coffee Shop  0.06


----Church and Welle

#### Put the results in a dataframe

First, let's write a function to sort the venues in descending order.

In [24]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

And create a new dataframe and display the top 10 venues for each neighborhood.

In [25]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
3                                 Central Bay Street           Coffee Shop   
4                                           Christie         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery          Cocktail Bar              Beer Bar   
1        Breakfast Spot                Bakery           Coffee Shop   
2       Airport Service      Airport Terminal               Airport   
3        Sandwich Place                  Café    Italian Restaurant   
4                  Café                  Park            Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0             Cheese Shop              Pharmacy    Seafood Restaurant   
1  Furniture / Home Store         Burrito Place               Stadium   
2                     Bar       Harbor / Marina                 Plane   
3             Salad Place       Thai Restaurant       Bubble Tea Shop   
4              Baby Store    Italian Restaurant    Athletics & Sports   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Farmers Market            Restaurant          Shopping Mall  
1            Restaurant    Italian Restaurant                    Bar  
2   Rental Car Location      Sculpture Garden               Boutique  
3          Burger Joint             Juice Bar      Korean Restaurant  
4             Nightclub           Coffee Shop            Candy Store

#### Cluster Neighborhoods

Finaly let's cluster the results. We will create 5 clusters.

In [26]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:20] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1, 1, 0])

In [27]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_toronto

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() 

Postal Code                 Borough                    Neighborhood  \
0         M4E            East Toronto                     The Beaches   
1         M4J  East York/East Toronto              The Danforth  East   
2         M4K            East Toronto    The Danforth West, Riverdale   
3         M4L            East Toronto  India Bazaar, The Beaches West   
4         M4M            East Toronto                 Studio District   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.676357 -79.293031               1                 Trail   
1  43.685347 -79.338106               0           Pizza Place   
2  43.679557 -79.352188               1      Greek Restaurant   
3  43.668999 -79.315572               1                  Park   
4  43.659526 -79.340923               1           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0     Health Food Store                   Pub            Donut Shop   
1          Intersection                  Park     Convenience Store   
2           Coffee Shop    Italian Restaurant        Ice Cream Shop   
3        Sandwich Place  Fast Food Restaurant             Pet Store   
4   American Restaurant                Bakery               Brewery   

  5th Most Common Venue 6th Most Common Venue   7th Most Common Venue  \
0                 Diner        Discount Store     Distribution Center   
1  Ethiopian Restaurant           Escape Room       Electronics Store   
2             Bookstore            Restaurant  Furniture / Home Store   
3                   Pub            Board Shop              Steakhouse   
4                  Café             Gastropub             Yoga Studio   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                      Dog Run      Doner Restaurant            Yoga Studio  
1  Eastern European Restaurant   Dumpling Restaurant             Donut Shop  
2                  Yoga Studio     Indian Restaurant   Caribbean Restaurant  
3           Italian Restaurant     Fish & Chips Shop       Sushi Restaurant  
4                  Fish Market             Pet Store                   Park

#### Let's visualize what we got

In [28]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Cluster 1

In [29]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
1  The Danforth  East               0           Pizza Place   
5       Lawrence Park               0      Business Service   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1          Intersection                  Park     Convenience Store   
5                  Park           Swim School              Bus Line   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1  Ethiopian Restaurant           Escape Room     Electronics Store   
5           Yoga Studio  Ethiopian Restaurant           Escape Room   

         8th Most Common Venue        9th Most Common Venue  \
1  Eastern European Restaurant          Dumpling Restaurant   
5            Electronics Store  Eastern European Restaurant   

  10th Most Common Venue  
1             Donut Shop  
5    Dumpling Restaurant

#### Cluster 2

In [30]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
0                                         The Beaches               1   
2                        The Danforth West, Riverdale               1   
3                      India Bazaar, The Beaches West               1   
4                                     Studio District               1   
6                                    Davisville North               1   
7                                  North Toronto West               1   
8                                          Davisville               1   
10  Summerhill West, Rathnelly, South Hill, Forest...               1   
12                        St. James Town, Cabbagetown               1   
13                               Church and Wellesley               1   
14                          Regent Park, Harbourfront               1   
15                           Garden District, Ryerson               1   
16                                     St. James Town               1   
17                                        Berczy Park               1   
18                                 Central Bay Street               1   
19                           Richmond, Adelaide, King               1   
20  Harbourfront East, Union Station, Toronto Islands               1   
21           Toronto Dominion Centre, Design Exchange               1   
22                     Commerce Court, Victoria Hotel               1   
25                The Annex, North Midtown, Yorkville               1   
26                     University of Toronto, Harbord               1   
27          Kensington Market, Chinatown, Grange Park               1   
28  CN Tower, King and Spadina, Railway Lands, Har...               1   
29                                     Enclave of M5E               1   
30             First Canadian Place, Underground city               1   
31                                           Christie               1   
32                       Dufferin, Dovercourt Village               1   
33                           Little Portugal, Trinity               1   
34       Brockton, Parkdale Village, Exhibition Place               1   
35                      High Park, The Junction South               1   
36                             Parkdale, Roncesvalles               1   
37                                 Runnymede, Swansea               1   
38                                     Enclave of M4L               1   

   1st Most Common Venue  2nd Most Common Venue 3rd Most Common Venue  \
0                  Trail      Health Food Store                   Pub   
2       Greek Restaurant            Coffee Shop    Italian Restaurant   
3                   Park         Sandwich Place  Fast Food Restaurant   
4            Coffee Shop    American Restaurant                Bakery   
6       Department Store                   Park        Breakfast Spot   
7         Clothing Store            Coffee Shop           Yoga Studio   
8           Dessert Shop         Sandwich Place           Pizza Place   
10           Coffee Shop            Supermarket                   Pub   
12           Coffee Shop                 Bakery            Restaurant   
13           Coffee Shop       Sushi Restaurant   Japanese Restaurant   
14           Coffee Shop                    Pub                Bakery   
15        Clothing Store            Coffee Shop       Bubble Tea Shop   
16           Coffee Shop                   Café        Cosmetics Shop   
17           Coffee Shop                 Bakery          Cocktail Bar   
18           Coffee Shop         Sandwich Place                  Café   
19           Coffee Shop                   Café            Restaurant   
20           Coffee Shop               Aquarium                 Hotel   
21           Coffee Shop                  Hotel                  Café   
22           Coffee Shop             Restaurant                 Hotel   
25                  Café         Sandwich Place           Coffee Shop   
26                  Café 

#### Cluster 3

In [31]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
9  Moore Park, Summerhill East               2           Summer Camp   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
9          Tennis Court           Yoga Studio          Dessert Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
9           Event Space  Ethiopian Restaurant           Escape Room   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
9     Electronics Store  Eastern European Restaurant    Dumpling Restaurant

#### Cluster 4

In [32]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
23     Roselawn               3          Home Service                Garden   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
23           Yoga Studio          Dessert Shop           Event Space   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
23  Ethiopian Restaurant           Escape Room     Electronics Store   

          9th Most Common Venue 10th Most Common Venue  
23  Eastern European Restaurant    Dumpling Restaurant

#### Cluster 5

In [33]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
11                  Rosedale               4                  Park   
24  Forest Hill North & West               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
11            Playground                 Trail           Yoga Studio   
24                 Trail         Jewelry Store      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
11      Department Store           Escape Room     Electronics Store   
24           Yoga Studio                 Diner  Ethiopian Restaurant   

          8th Most Common Venue 9th Most Common Venue  \
11  Eastern European Restaurant   Dumpling Restaurant   
24                  Escape Room     Electronics Store   

         10th Most Common Venue  
11                   Donut Shop  
24  Eastern European Restaurant

#### Short summary

We can see that Cluster 2 has the most members. Other clusters 2 or 1 member. I tried with different number of clusters but the figures were the same, cluster 2 had the most members. It could be interesting to run clustering on all boroughs and neighborhoods of Toronto.